src/olympia/addons/utils.py

```diff
@@ -247,3 +259,130 @@ def get_error_message(self):
         except IndexError:
             msg = None
         return msg
+
+
+class PermissionEnablerCreator:
+    """Helper class to create new permission enabler add-ons and versions.
+
+    Assumes parameters have already been validated beforehand."""
+
+    SITE_PERMISSIONS = {
+        'webmidi': 'WebMIDI',
+    }
+    MIN_VERSION = '96.0'
+
+    def __init__(self, *, user, remote_addr, install_origins, site_permissions):
+        self.user = user
+        self.remote_addr = remote_addr
+        self.install_origins = install_origins
+        self.site_permissions = site_permissions
+
+    def _create_manifest(self, version_number):
+
+        hostnames = ', '.join(
+            [urlparse(origin).netloc for origin in self.install_origins]
+        )
+        permissions = ', '.join(
+            self.SITE_PERMISSIONS.get(permission, permission)
+            for permission in self.site_permissions
+        )
+        # FIXME: https://github.com/mozilla/addons-server/issues/18421 to
+        # generate translations for the name & description (we'll need
+        # __MSG_extensionName__and _locales/ folder etc). At the moment the
+        # gettext call is just here to prepare for the future and allow
+        # translators to work on translations, but we're always generating in
+        # english.
+        with translation.override('en-US'):
+            manifest_data = {
+                'manifest_version': 2,
+                'version': version_number,
+                'name': gettext('Site permissions for {hostnames}').format(
+                    hostnames=hostnames
+                ),
+                'description': gettext(
+                    'This add-on provides {hostnames} with access to the following '
+                    'protected DOM APIs: {permissions}.'
+                ).format(hostnames=hostnames, permissions=permissions),
+                'install_origins': self.install_origins,
+                'site_permissions': self.site_permissions,
+                'browser_specific_settings': {
+                    'gecko': {'strict_min_version': self.MIN_VERSION}
+                },
+            }
+        return manifest_data
+
+    def _create_zipfile(self, manifest_data):
+        # Create the xpi containing the manifest, in memory.
+        raw_buffer = BytesIO()
+        with zipfile.ZipFile(raw_buffer, 'w') as zip_file:
+            zip_file.writestr('manifest.json', json.dumps(manifest_data, indent=2))
+        raw_buffer.seek(0)
+        filename = 'automatic.xpi'
```

Comment:
a more descriptive file name? (containing hostname or permission, etc)

Reply:
Ultimately it doesn't really matter that much as it's all in memory, it won't get written to the filesystem with that name. I just need _a_ name so that I can use a `File` object, which I need because that's the kind of object `parse_addon()` expects.